### PR TITLE
Install with no arguments

### PR DESCRIPTION
**Pull request by gigabo**
Use temporary package.json files.

This is groundwork for:
- Pluggable npm clients (yarn).
- An `asini upgrade` command.

### CODE REVIEW

File: src/NpmUtilities.js [hunk 1]
```diff
@@ -7,18 +9,66 @@ import semver from "semver";
...
+        }, {})
+      });
+
+      // Write out our temporary cooked up package.json and then install.
+      FileSystemUtilities.writeFile(packageJson, tempJson, (err) => {
+
+        // We have a few housekeeping tasks to take care of whether we succeed or fail.
+        const done = (err) => {
+          cleanup();
+          unregister();
```
Comment (rygine) [inline src/NpmUtilities.js:53]: is `cleanup()` not guaranteed to run in `unregister()`?

File: src/NpmUtilities.js [hunk 2]
```diff
@@ -7,18 +9,66 @@ import semver from "semver";
...
+      });
+    });
+  }
+
+  // Take a dep like "foo@^1.0.0".
+  // Return a tuple like ["foo", "^1.0.0"].
+  // Handles scoped packages.
+  // Returns undefined for version if none specified.
+  static splitVersion(dep) {
+    return dep.match(/^(@?[^@]+)(?:@(.+))?/).slice(1, 3);
```
Comment (rygine) [inline src/NpmUtilities.js:71]: i just love regexs 😖

**Comment by rygine** [comment]
so no more `npm install foo@1.x bar@3.2.2 quu@blah`, right?

edit: sorry, this was slightly rhetorical. i see that this is the case. just `npm install` is run. 👍

File: src/NpmUtilities.js (hunk 1 above)
Comment (gigabo) [reply src/NpmUtilities.js:53]: `unregister()` unschedules the `onExit()` cleanup, which we only need if we wind up exiting before this `done()` function is called.

File: src/NpmUtilities.js (hunk 2 above)
Comment (gigabo) [reply src/NpmUtilities.js:71]: Yeah, I originally used a simple `.split("@")`, but then I remembered scoped packages. 😖

**Comment by gigabo** [comment]
> so no more npm install foo@1.x bar@3.2.2 quu@blah, right?

Right!  This is important for `yarn`, which won't install anything that's not in the package.json.

**Pull request merged**